### PR TITLE
Allow cluster to start when only n/2 servers are up

### LINKED ISSUE #3234: erasure: Replace n/2 disks, new servers do not come up
## Steps to Reproduce (for bugs)
1. run 2:2 DXL
2. bring down minio3 and minio4 (i.e 3rd and 4th server)
3. rm -rf /export/disk{3..4} (i.e remove the down servers disks)
4. mkdir /export/disk{3..4}
5. start minio3 and minio4

minio 3 and 4 does not come up and prints eternally:
`Initializing data volume. Waiting for minimum 3 servers to come online.`

### PR DESCRIPTION
**Pull request by donatello**
## Description

Fixes #3234.

Relaxes the quorum requirement to start the object layer, and skips
quick-healing at start-up (as no write quorum is present).

This fix allows reads to work after start-up, but it does hang on writes. The hang is because no write quorum servers are available and the lock interface currently does not have a possibility of failing/returning an error. In my view, this is a separate issue - #3185 ought to be re-opened.

## Motivation and Context

Toward resolving #3234. 

## How Has This Been Tested?

Manually.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Review by krisis — APPROVED** [review]

File: cmd/prepare-storage.go [hunk 1]
```diff
@@ -150,29 +151,35 @@ func prepForInitXL(firstDisk bool, sErrs []error, diskCount int) InitActions {
...
+		// Only the first server formats an uninitialized
+		// setup, others wait for notification.
+		if firstDisk {
 			return FormatDisks
 		}
 		return WaitForFormatting
 	}
 
-	// Total disks unformatted are in quorum verify if we have some offline disks.
+	// Total disks unformatted are in quorum verify if we have
```
Comment (harshavardhana) [inline cmd/prepare-storage.go]: These are not related changes to PR, can we avoid it?

File: cmd/prepare-storage.go [hunk 2]
```diff
@@ -150,29 +151,35 @@ func prepForInitXL(firstDisk bool, sErrs []error, diskCount int) InitActions {
...
-	if disksFormatted >= quorum {
+	// Already formatted and having read-quorum, proceed to
+	// initialization of object layer.
+	if disksFormatted >= readQuorum {
 		if disksFormatted+disksOffline == diskCount {
 			return InitObjectLayer
 		}
-		// Some of the formatted disks are possibly corrupted or unformatted, heal them.
+		// Some of the formatted disks are possibly corrupted
+		// or unformatted, heal them.
```
Comment (harshavardhana) [inline cmd/prepare-storage.go]: If we still return `WaitForHeal` should the caller proceed to heal? under ReadQuorum?

File: cmd/xl-v1.go [hunk 3]
```diff
@@ -155,6 +155,21 @@ func newXLObjects(storageDisks []StorageAPI) (ObjectLayer, error) {
 	xl.readQuorum = readQuorum
 	xl.writeQuorum = writeQuorum
 
+	// If the number of offline servers is equal to the readQuorum
+	// (i.e. the number of online servers also equals the
+	// readQuorum), we cannot perform quick-heal (no
+	// write-quorum). However reads may still be possible, so we
+	// skip quick-heal in this case, and continue.
+	offlineCount := 0
```
Comment (harshavardhana) [inline cmd/xl-v1.go]: perhaps you can use `len(newStorageDisks) - diskCount(newStorageDisks)`  or create a function similar to `diskCount` and rename it as `onlineDiskCount` and this one has `offlineDiskCount`

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: cmd/prepare-storage.go (hunk 1 above)
Comment (donatello) [reply cmd/prepare-storage.go]: Done.

File: cmd/xl-v1.go (hunk 3 above)
Comment (donatello) [reply cmd/xl-v1.go]: Done.

File: cmd/prepare-storage.go (hunk 2 above)
Comment (donatello) [reply cmd/prepare-storage.go]: From the current implementation in `healFormatXL`, this seems ok as no quorum appears to be required in it - I am not sure if that is entirely correct, but in this case, there seems to be no harm in allowing to heal format.json on a unformatted/corrupted disk if there are readQuorum other formatted disks. E.g.: in a fully working 8-node 8-disk cluster, if 4-disks are replaced, there is no data-loss (due to erasure-coding) and healing should be allowed.

**Comment by donatello** [comment]
@harshavardhana Your comments are addressed - please check now.

**Review by harshavardhana — APPROVED** [review]

**Comment by harshavardhana** [comment]
Ping @krishnasrinivas  can you review this once ?

File: cmd/prepare-storage.go [hunk 4]
```diff
@@ -132,13 +132,14 @@ func prepForInitXL(firstDisk bool, sErrs []error, diskCount int) InitActions {
...
 	quorum := diskCount/2 + 1
+	readQuorum := diskCount / 2
 	disksOffline := errMap[errDiskNotFound]
 	disksFormatted := errMap[nil]
 	disksUnformatted := errMap[errUnformattedDisk]
 	disksCorrupted := errMap[errCorruptedFormat]
 
 	// No Quorum lots of offline disks, wait for quorum.
-	if disksOffline >= quorum {
+	if disksOffline > readQuorum {
```
Comment (krishnasrinivas) [inline cmd/prepare-storage.go:142]: Not related to this commit. But when deciding on quorum we have to consider online-disks rather than offline-disks.

Because when data != parity this comparison will be incorrect.

For ex. consider 6-data and 2-parity. readQuorum will be 6. If 3 disks are down, the statement becomes:
```
if 3 > 6 {
    return WaitForQuorum
}
```
Which is incorrect because we should have returned WaitForQuorum.

If we talk in terms of online-disks, it will be like this (which is also more readable):
```
if disksOnline < readQuorum {
    return WaitForQuorum
}
```
which will work as intended.

If/when we make data/parity configurable we will have to fix these things.

**Review by krishnasrinivas — APPROVED** [review]

File: cmd/prepare-storage.go (hunk 4 above)
Comment (harshavardhana) [reply cmd/prepare-storage.go:142]: Can this be captured in a bug @krishnasrinivas ?

Comment (krishnasrinivas) [reply cmd/prepare-storage.go:142]: @harshavardhana opened an umbrella issue https://github.com/minio/minio/issues/4079
